Add Mac OS tests to Github Action

### Current problem

The tests suite is not run on Mac so the tests are failing locally for mac developpers.

### Desired solution

Run Mac tests in Github Actions.

### Additional context

See discussion in https://github.com/PyCQA/pylint/pull/4753#issuecomment-887819516

I reported this as running the test suite on my local machine produced different errors than those reported by the GitHub actions.

I have added the report. Perhaps this is due to something I messed up in my local environment, but I did not fully understand why these errors were only thrown on my local machine.

```
GLOB sdist-make: /Users/daniel/DocumentenLaptop/Programming/Github/pylint/setup.py
py39 recreate: /Users/daniel/DocumentenLaptop/Programming/Github/pylint/.tox/py39
py39 installdeps: -r/Users/daniel/DocumentenLaptop/Programming/Github/pylint/requirements_test.txt
py39 inst: /Users/daniel/DocumentenLaptop/Programming/Github/pylint/.tox/.tmp/package/3/pylint-2.9.6.dev0.zip
py39 installed: appdirs==1.4.4,astroid==2.6.5,attrs==19.3.0,backports.entry-points-selectable==1.1.0,black==21.7b0,certifi==2021.5.30,cfgv==3.3.0,charset-normalizer==2.0.3,cli-ui==0.10.3,click==8.0.1,colorama==0.4.4,contextlib2==21.6.0,coverage==5.5,coveralls==3.2.0,distlib==0.3.2,docopt==0.6.2,execnet==1.9.0,filelock==3.0.12,flake8==3.9.2,gprof2dot==2021.2.21,identify==2.2.11,idna==3.2,iniconfig==1.1.1,isort==5.9.2,lazy-object-proxy==1.6.0,mccabe==0.6.1,mypy==0.910,mypy-extensions==0.4.3,nodeenv==1.6.0,packaging==21.0,pathspec==0.9.0,platformdirs==2.1.0,pluggy==0.13.1,pre-commit==2.13.0,py==1.10.0,py-cpuinfo==8.0.0,pycodestyle==2.7.0,pyenchant==3.2.1,pyflakes==2.3.1,pylint @ file:///Users/daniel/DocumentenLaptop/Programming/Github/pylint/.tox/.tmp/package/3/pylint-2.9.6.dev0.zip,pyparsing==2.4.7,pytest==6.2.4,pytest-benchmark==3.4.1,pytest-cov==2.12.1,pytest-forked==1.3.0,pytest-profiling==1.7.0,pytest-xdist==2.3.0,PyYAML==5.4.1,regex==2021.7.6,requests==2.26.0,schema==0.7.4,six==1.16.0,tabulate==0.8.9,tbump==6.3.2,toml==0.10.2,tomli==1.1.0,tomlkit==0.7.2,types-pkg-resources==0.1.3,types-toml==0.1.3,typing-extensions==3.10.0.0,Unidecode==1.2.0,urllib3==1.26.6,virtualenv==20.6.0,wrapt==1.12.1
py39 run-test-pre: PYTHONHASHSEED='2685728459'
py39 run-test: commands[0] | pytest --benchmark-disable /Users/daniel/DocumentenLaptop/Programming/Github/pylint/tests/
============================= test session starts ==============================
platform darwin -- Python 3.9.6, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
cachedir: .tox/py39/.pytest_cache
benchmark: 3.4.1 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /Users/daniel/DocumentenLaptop/Programming/Github/pylint, configfile: setup.cfg, testpaths: tests
plugins: cov-2.12.1, benchmark-3.4.1, xdist-2.3.0, profiling-1.7.0, forked-1.3.0
collected 1681 items / 204 deselected / 1477 selected

tests/test_check_parallel.py ...........................                 [  1%]
tests/test_config.py ....                                                [  2%]
tests/test_epylint.py ..                                                 [  2%]
tests/test_func.py ............                                          [  3%]
tests/test_functional.py F.............................................. [  6%]
........F.........................s....................................s [ 11%]
......................................................FF.....F.......... [ 15%]
.............................................s.s.ss....s................ [ 20%]
........s..........F....s...............................s............... [ 25%]
.F....F.................F...............................s............... [ 30%]
..........................ss.s.s.............F.F.s...................... [ 35%]
.....s....s.s.s................................                          [ 38%]
tests/test_import_graph.py ......                                        [ 39%]
tests/test_numversion.py ...........                                     [ 39%]
tests/test_pragma_parser.py ............                                 [ 40%]
tests/test_pylint_runners.py ....                                        [ 40%]
tests/test_regr.py ..................                                    [ 42%]
tests/test_self.py ..................................................... [ 45%]
....................................................                     [ 49%]
tests/unittest_config.py ......                                          [ 49%]
tests/unittest_pyreverse_diadefs.py .........                            [ 50%]
tests/unittest_pyreverse_inspector.py ........                           [ 50%]
tests/unittest_pyreverse_writer.py ..................                    [ 51%]
tests/unittest_reporters_json.py .                                       [ 52%]
tests/unittest_reporting.py .....                                        [ 52%]
tests/benchmark/test_baseline_benchmarks.py ...........                  [ 53%]
tests/checkers/unittest_base.py .............s..........s                [ 54%]
tests/checkers/unittest_classes.py .........                             [ 55%]
tests/checkers/unittest_deprecated.py ........................           [ 57%]
tests/checkers/unittest_exceptions.py ..                                 [ 57%]
tests/checkers/unittest_format.py ..............                         [ 58%]
tests/checkers/unittest_imports.py ......                                [ 58%]
tests/checkers/unittest_logging.py ........                              [ 59%]
tests/checkers/unittest_misc.py ..........                               [ 59%]
tests/checkers/unittest_python3.py ..................................... [ 62%]
.....................................................                    [ 65%]
tests/checkers/unittest_refactoring.py .                                 [ 65%]
tests/checkers/unittest_similar.py ...............                       [ 66%]
tests/checkers/unittest_spelling.py ssssssssssssssssssssssssssssssssss   [ 69%]
tests/checkers/unittest_stdlib.py ......                                 [ 69%]
tests/checkers/unittest_strings.py ...                                   [ 69%]
tests/checkers/unittest_typecheck.py ....................                [ 71%]
tests/checkers/unittest_utils.py ..................................      [ 73%]
tests/checkers/unittest_variables.py .....................               [ 74%]
tests/extensions/test_bad_builtin.py .                                   [ 75%]
tests/extensions/test_broad_try_clause.py .                              [ 75%]
tests/extensions/test_check_docs.py .................................... [ 77%]
........................................................................ [ 82%]
..............                                                           [ 83%]
tests/extensions/test_check_docs_utils.py ..............                 [ 84%]
tests/extensions/test_check_mccabe.py ..                                 [ 84%]
tests/extensions/test_check_raise_docs.py .............................. [ 86%]
................                                                         [ 87%]
tests/extensions/test_check_return_docs.py ............................. [ 89%]
...........                                                              [ 90%]
tests/extensions/test_check_yields_docs.py ............................  [ 92%]
tests/extensions/test_comparetozero.py .                                 [ 92%]
tests/extensions/test_confusing_elif.py ........                         [ 92%]
tests/extensions/test_docstyle.py .                                      [ 92%]
tests/extensions/test_elseif_used.py .                                   [ 92%]
tests/extensions/test_empty_comment.py .                                 [ 92%]
tests/extensions/test_emptystring.py .                                   [ 93%]
tests/extensions/test_overlapping_exceptions.py .F                       [ 93%]
tests/extensions/test_redefined.py .                                     [ 93%]
tests/functional/r/redundant_unittest_assert.py ss                       [ 93%]
tests/lint/unittest_expand_modules.py ....                               [ 93%]
tests/lint/unittest_lint.py ............................................ [ 96%]
.......                                                                  [ 97%]
tests/message/unittest_message.py .                                      [ 97%]
tests/message/unittest_message_definition.py ......                      [ 97%]
tests/message/unittest_message_definition_store.py .................     [ 98%]
tests/message/unittest_message_id_store.py .......                       [ 99%]
tests/profile/test_profile_against_externals.py s                        [ 99%]
tests/testutils/test_output_line.py .......                              [ 99%]
tests/utils/unittest_ast_walker.py ..                                    [ 99%]
tests/utils/unittest_utils.py ..                                         [100%]

=================================== FAILURES ===================================
___________________ test_functional[import_outside_toplevel] ___________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x10ca69d00>

    def runTest(self):
>       self._runTest()
E       AssertionError: Wrong results for file "import_outside_toplevel":
E       
E       Unexpected in testdata:
E         42: no-name-in-module

.tox/py39/lib/python3.9/site-packages/pylint/testutils/lint_module_test.py:80: AssertionError
______________ test_functional[regression_1326_crash_uninferable] ______________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x1118d3f40>

    def runTest(self):
>       self._runTest()
E       AssertionError: Wrong results for file "regression_1326_crash_uninferable":
E       
E       Unexpected in testdata:
E          5: no-member

.tox/py39/lib/python3.9/site-packages/pylint/testutils/lint_module_test.py:80: AssertionError
__________________ test_functional[invalid_exceptions_raised] __________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x113a0b520>

    def runTest(self):
>       self._runTest()
E       AssertionError: Wrong results for file "invalid_exceptions_raised":
E       
E       Unexpected in testdata:
E         22: c-extension-no-member

.tox/py39/lib/python3.9/site-packages/pylint/testutils/lint_module_test.py:80: AssertionError
__________________ test_functional[invalid_exceptions_caught] __________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x114c4e760>

    def runTest(self):
>       self._runTest()
E       AssertionError: Wrong results for file "invalid_exceptions_caught":
E       
E       Unexpected in testdata:
E         19: no-member
E         98: c-extension-no-member

.tox/py39/lib/python3.9/site-packages/pylint/testutils/lint_module_test.py:80: AssertionError
____________________ test_functional[inconsistent_returns] _____________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x112f174c0>

    def runTest(self):
>       self._runTest()
E       AssertionError: Wrong results for file "inconsistent_returns":
E       
E       Unexpected in testdata:
E         10: c-extension-no-member
E         17: c-extension-no-member
E        162: c-extension-no-member
E        167: c-extension-no-member

.tox/py39/lib/python3.9/site-packages/pylint/testutils/lint_module_test.py:80: AssertionError
_____________________ test_functional[socketerror_import] ______________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x118105d00>

    def runTest(self):
>       self._runTest()
E       AssertionError: Wrong results for file "socketerror_import":
E       
E       Unexpected in testdata:
E          3: no-name-in-module

.tox/py39/lib/python3.9/site-packages/pylint/testutils/lint_module_test.py:80: AssertionError
_____________________ test_functional[function_redefined] ______________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x11801d370>

    def runTest(self):
>       self._runTest()
E       AssertionError: Wrong results for file "function_redefined":
E       
E       Unexpected in testdata:
E         85: no-name-in-module

.tox/py39/lib/python3.9/site-packages/pylint/testutils/lint_module_test.py:80: AssertionError
_______________ test_functional[f_string_without_interpolation] ________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x114277580>

    def runTest(self):
>       self._runTest()
E       AssertionError: Wrong results for file "f_string_without_interpolation":
E       
E       Unexpected in testdata:
E          7: c-extension-no-member

.tox/py39/lib/python3.9/site-packages/pylint/testutils/lint_module_test.py:80: AssertionError
________________________ test_functional[member_checks] ________________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x1182b4eb0>

    def runTest(self):
>       self._runTest()
E       AssertionError: Wrong results for file "member_checks":
E       
E       Unexpected in testdata:
E        151: no-member

.tox/py39/lib/python3.9/site-packages/pylint/testutils/lint_module_test.py:80: AssertionError
_____________________ test_functional[builtin_module_test] _____________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x11ac0fbb0>

    def runTest(self):
>       self._runTest()
E       AssertionError: Wrong results for file "builtin_module_test":
E       
E       Unexpected in testdata:
E          4: no-name-in-module

.tox/py39/lib/python3.9/site-packages/pylint/testutils/lint_module_test.py:80: AssertionError
____________________ test_functional[bad_exception_context] ____________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x11aba2bb0>

    def runTest(self):
>       self._runTest()
E       AssertionError: Wrong results for file "bad_exception_context":
E       
E       Unexpected in testdata:
E         19: no-member

.tox/py39/lib/python3.9/site-packages/pylint/testutils/lint_module_test.py:80: AssertionError
_______________________ test_overlapping_exceptions_py33 _______________________

linter = Checker 'master' (responsible for 'F0001', 'F0002', 'F0010', 'I0001', 'I0010', 'I0011', 'I0013', 'I0020', 'I0021', 'I0022', 'E0001', 'E0011', 'E0012', 'E0013')

>   ???
E   AssertionError: assert 'E1101' == 'W0714'
E     - W0714
E     + E1101

/Users/daniel/Documents/Programming/Github/DanielNoord/pylint/tests/extensions/test_overlapping_exceptions.py:84: AssertionError
=========================== short test summary info ============================
FAILED tests/test_functional.py::test_functional[import_outside_toplevel] - A...
FAILED tests/test_functional.py::test_functional[regression_1326_crash_uninferable]
FAILED tests/test_functional.py::test_functional[invalid_exceptions_raised]
FAILED tests/test_functional.py::test_functional[invalid_exceptions_caught]
FAILED tests/test_functional.py::test_functional[inconsistent_returns] - Asse...
FAILED tests/test_functional.py::test_functional[socketerror_import] - Assert...
FAILED tests/test_functional.py::test_functional[function_redefined] - Assert...
FAILED tests/test_functional.py::test_functional[f_string_without_interpolation]
FAILED tests/test_functional.py::test_functional[member_checks] - AssertionEr...
FAILED tests/test_functional.py::test_functional[builtin_module_test] - Asser...
FAILED tests/test_functional.py::test_functional[bad_exception_context] - Ass...
FAILED tests/extensions/test_overlapping_exceptions.py::test_overlapping_exceptions_py33
=== 12 failed, 1406 passed, 59 skipped, 204 deselected in 144.66s (0:02:24) ====
ERROR: InvocationError for command /Users/daniel/DocumentenLaptop/Programming/Github/pylint/.tox/py39/bin/pytest --benchmark-disable tests (exited with code 1)
___________________________________ summary ____________________________________
ERROR:   py39: commands failed

```

--- a/.github/workflows/tests.yaml
+++ b/.github/workflows/tests.yaml
@@ -58,7 +58,7 @@ jobs:
       - name: Run pytest
         run: |
           . venv/bin/activate
-          pytest -vv --durations=20 --benchmark-disable --cov --cov-report= tests/
+          pytest --benchmark-disable --cov --cov-report= tests/
       - name: Run functional tests with minimal messages config
         run: |
           . venv/bin/activate
@@ -213,7 +213,7 @@ jobs:
       - name: Run pytest
         run: |
           . venv\\Scripts\\activate
-          pytest -vv --durations=20 --benchmark-disable tests/
+          pytest --benchmark-disable tests/
 
   tests-macos:
     name: run / ${{ matrix.python-version }} / macOS
@@ -223,7 +223,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        # We only run on the latest supported version on Mac
+        # We only run on the oldest supported version on Mac
         python-version: [3.7]
     steps:
       - name: Check out code from GitHub
@@ -259,7 +259,7 @@ jobs:
       - name: Run pytest
         run: |
           . venv/bin/activate
-          pytest -vv --durations=20 --benchmark-disable tests/
+          pytest --benchmark-disable tests/
 
   tests-pypy:
     name: run / ${{ matrix.python-version }} / Linux
@@ -303,4 +303,4 @@ jobs:
       - name: Run pytest
         run: |
           . venv/bin/activate
-          pytest -vv --durations=20 --benchmark-disable tests/
+          pytest --benchmark-disable tests/

--- a/.github/workflows/tests.yaml
+++ b/.github/workflows/tests.yaml
@@ -58,7 +58,7 @@ jobs:
       - name: Run pytest
         run: |
           . venv/bin/activate
-          pytest --benchmark-disable --cov --cov-report= tests/
+          pytest --durations=10 --benchmark-disable --cov --cov-report= tests/
       - name: Run functional tests with minimal messages config
         run: |
           . venv/bin/activate
@@ -213,7 +213,7 @@ jobs:
       - name: Run pytest
         run: |
           . venv\\Scripts\\activate
-          pytest --benchmark-disable tests/
+          pytest --durations=10 --benchmark-disable tests/
 
   tests-macos:
     name: run / ${{ matrix.python-version }} / macOS
@@ -259,7 +259,7 @@ jobs:
       - name: Run pytest
         run: |
           . venv/bin/activate
-          pytest --benchmark-disable tests/
+          pytest --durations=10 --benchmark-disable tests/
 
   tests-pypy:
     name: run / ${{ matrix.python-version }} / Linux
@@ -303,4 +303,4 @@ jobs:
       - name: Run pytest
         run: |
           . venv/bin/activate
-          pytest --benchmark-disable tests/
+          pytest --durations=10 --benchmark-disable tests/

--- a/.github/workflows/tests.yaml
+++ b/.github/workflows/tests.yaml
@@ -215,6 +215,52 @@ jobs:
           . venv\\Scripts\\activate
           pytest -vv --durations=20 --benchmark-disable tests/
 
+  tests-macos:
+    name: run / ${{ matrix.python-version }} / macOS
+    runs-on: macos-latest
+    timeout-minutes: 25
+    needs: tests-linux
+    strategy:
+      fail-fast: false
+      matrix:
+        # We only run on the latest supported version on Mac
+        python-version: [3.7]
+    steps:
+      - name: Check out code from GitHub
+        uses: actions/checkout@v3.0.2
+      - name: Set up Python ${{ matrix.python-version }}
+        id: python
+        uses: actions/setup-python@v3.1.2
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Generate partial Python venv restore key
+        id: generate-python-key
+        run: >-
+          echo "::set-output name=key::venv-${{ env.CACHE_VERSION }}-${{
+            hashFiles('setup.cfg', 'requirements_test_min.txt')
+          }}"
+      - name: Restore Python virtual environment
+        id: cache-venv
+        uses: actions/cache@v3.0.2
+        with:
+          path: venv
+          key: >-
+            ${{ runner.os }}-${{ steps.python.outputs.python-version }}-${{
+            steps.generate-python-key.outputs.key }}
+          restore-keys: |
+            ${{ runner.os }}-${{ steps.python.outputs.python-version }}-venv-${{ env.CACHE_VERSION }}-
+      - name: Create Python virtual environment
+        if: steps.cache-venv.outputs.cache-hit != 'true'
+        run: |
+          python -m venv venv
+          . venv/bin/activate
+          python -m pip install -U pip setuptools wheel
+          pip install -U -r requirements_test.txt
+      - name: Run pytest
+        run: |
+          . venv/bin/activate
+          pytest -vv --durations=20 --benchmark-disable tests/
+
   tests-pypy:
     name: run / ${{ matrix.python-version }} / Linux
     runs-on: ubuntu-latest

